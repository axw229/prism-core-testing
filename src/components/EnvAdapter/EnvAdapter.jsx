// @flow
/* global CustomEvent */
import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import at from 'lodash/at'

export const getStoreValues = (store: Object): Object => {
  return {
    chosenColor: at(store, 'colors.chosenColor')[0],
    colorWallActive: at(store, 'colors.colorWallActive')[0],
    emitColor: at(store, 'colors.emitColor')[0],
    family: at(store, 'colors.family')[0],
    query: at(store, 'colors.search.query')[0],
    section: at(store, 'colors.section')[0]
  }
}

// TODO: hydrate EVENT_NAME_PREFIX w/ config value for customizable event catching
export const EVENT_NAME_PREFIX = 'swPrism.broadcastChange'

export type Props = {
  id?: string,
  className?: string,
  store: {
    chosenColor?: any,
    emitColor?: any,
    colorWallActive?: any,
    family?: any,
    section?: any,
    query?: any
  }
}

const defaultProps: Props = {
  store: {}
}

export class EnvAdapter extends PureComponent<Props> {
  ref: ?RefObject = void (0)

  static defaultProps: Props = defaultProps

  constructor (props: Props) {
    super()

    this.ref = React.createRef()
    this.emit(props)
  }

  render () {
    const { id, className } = this.props

    return (
      <script id={id} className={className} ref={this.ref} />
    )
  }

  componentDidUpdate (prevProps: Props) {
    this.emit(this.props, prevProps)
  }

  emit = (is: Props, was: Props = defaultProps) => {
    const { store: prevStore } = was
    const { store } = is

    if (store && prevStore) {
      const keys = Object.keys(store)

      keys.forEach(key => {
        if (store[key] !== prevStore[key]) {
          const event = new CustomEvent(`${EVENT_NAME_PREFIX}.${key}`, {
            bubbles: true,
            detail: store[key]
          })

          if (this.ref && this.ref.current) {
            this.ref.current.dispatchEvent(event)
          }
        }
      })
    }
  }
}

const mapStateToProps = (state, props) => {
  return {
    // the "store" prop relates to data in redux
    store: getStoreValues(state)
  }
}

export default connect(mapStateToProps)(EnvAdapter)
