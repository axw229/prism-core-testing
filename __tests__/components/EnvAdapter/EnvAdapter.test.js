/* eslint-env jest */
// @flow
import React from 'react'
import { mount } from 'enzyme'
import { EnvAdapter, type Props, EVENT_NAME_PREFIX, getStoreValues } from '../../../src/components/EnvAdapter/EnvAdapter'

const getComponent = (props: Props) => {
  const defaultProps = {}
  const newProps = {
    ...defaultProps,
    ...props
  }

  return mount(<EnvAdapter {...newProps} />)
}

describe('EnvAdapter', () => {
  let basicEnvAdapter

  beforeAll(() => {
    if (!basicEnvAdapter) {
      basicEnvAdapter = getComponent()
    }
  })

  it('should match snapshot', () => {
    expect(basicEnvAdapter).toMatchSnapshot()
  })

  it('should emit something when its watched properties are changed', (done) => {
    const comp = getComponent()
    const node = comp.getDOMNode()
    // get all watched properties of EnvAdapter
    const properties = Object.keys(getStoreValues())
    // generate random values paired to each property
    const values = properties.map((val, i) => `${val} test value`)

    // loop over all watched properties...
    properties.forEach((prop: string, i: number) => {
      const testValue = values[i]
      const testEvent = `${EVENT_NAME_PREFIX}.${prop}`

      // ... add an event listener for each property
      node.addEventListener(testEvent, (event) => {
        expect(event.detail).toEqual(testValue)
        done()
      })

      // and set each property in sequence to our test value
      comp.setProps({
        store: {
          [prop]: testValue
        }
      })
    })
  })
})
