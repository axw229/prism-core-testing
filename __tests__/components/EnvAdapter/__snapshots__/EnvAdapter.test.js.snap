// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvAdapter should match snapshot 1`] = `
<EnvAdapter
  store={Object {}}
>
  <script />
</EnvAdapter>
`;
